src/pages/FiveCoinPuzzle.js

```diff
@@ -0,0 +1,201 @@
+import "../styles.css";
+import React, { useEffect, useState } from "react";
+import toast, { Toaster } from "react-hot-toast";
+import ArrowKeysReact from "arrow-keys-react";
+import Button from "../components/Button";
+import Coin from "../components/Coin";
+
+export default function FiveCoinPuzzle() {
+  const numberOfCoinSlots = 17;
+  const startingArray = [0, 0, 0, 0, 0, 0, 1, 2, 1, 2, 1, 0, 0, 0, 0, 0, 0];
+  // player wins if either of these segments are present at any point in the array
+  const winningArraySegments = [
+    [2, 2, 1, 1, 1],
+    [1, 1, 1, 2, 2],
+  ];
+  const [coinTypeArray, setCoinTypeArray] = useState(startingArray);
+  const [firstSelectedCoin, setFirstSelectedCoin] = useState(-1);
+  const [secondSelectedCoin, setSecondSelectedCoin] = useState(-1);
+  const [moveCounter, setMoveCounter] = useState(0);
+  const [coinSelectText, setcoinSelectText] = useState(
+    "Click to select first coin"
+  );
+
+  const swapCoins = (array, index1, index2) => {
+    [array[index1], array[index2]] = [array[index2], array[index1]];
+    setCoinTypeArray(array);
+    selectCoin(index2);
+    checkCompletion(array);
+  };
+  const selectCoin = (coinNumber) => {
+    // cannot select empty coins
+    if (coinTypeArray[coinNumber] !== 0) {
+      if (firstSelectedCoin === -1) {
+        setFirstSelectedCoin(coinNumber);
+        setcoinSelectText("Click to select second coin");
+      } else if (
+        secondSelectedCoin === -1 &&
+        (coinNumber === firstSelectedCoin - 1 ||
+          coinNumber === firstSelectedCoin + 1) &&
+        coinTypeArray[coinNumber] !== coinTypeArray[firstSelectedCoin]
+      ) {
+        setSecondSelectedCoin(coinNumber);
+        setcoinSelectText(
+          `Use arrow keys to move coins
+          ESC to deselect coins`
+        );
+      } else return;
+    }
+  };
+  const areTwoCoinsSelected = () => {
```

Comment:
Functions within components get re-rendered whenever the component gets re-rendered. So for functions like this one that don't have logic intrinsically important to the components logic, it's nice to pull it out into either a separate file, or just at top of this file so it doesn't get rerendered all the time (since it doesn't change). You can then make it more generic:

```
const areBothCoinsSelected = (coinA, coinB) => {
  if (firstSelectedCoin === -1 || secondSelectedCoin === -1) {
      return false;
    }
    return true;
}
```
and then call:
```
areBothCoinsSelected(firstSelectedCoin, secondSelectedCoin)
```